packages/crypto/src/proof.ts

```diff
@@ -0,0 +1,254 @@
+import type { v1AppProof, v1BootProof } from "@turnkey/sdk-types";
+import { p256 } from "@noble/curves/p256";
+import { sha256 } from "@noble/hashes/sha2";
+import * as CBOR from "cbor-js";
+import * as x509 from "@peculiar/x509";
+import { Crypto as PeculiarCrypto } from "@peculiar/webcrypto";
+import { createHash, createPublicKey, createVerify } from "node:crypto";
+import { AWS_ROOT_CERT_PEM, AWS_ROOT_CERT_SHA256 } from "./constants";
+
+x509.cryptoProvider.set(new PeculiarCrypto());
+
+// verify goes through the following verification steps for an app proof & boot proof pair:
+//  - Verify app proof signature
+//  - Verify the boot proof
+//    - Attestation doc was signed by AWS
+//    - Attestation doc's `user_data` is the hash of the qos manifest
+//  - Verify the connection between the app proof & boot proof i.e. that the ephemeral key's match
+//
+// For more information, check out https://whitepaper.turnkey.com/foundations
+export async function verify(
+  appProof: v1AppProof,
+  bootProof: v1BootProof,
+): Promise<void> {
+  // 1. Verify App Proof
+  verifyAppProofSignature(appProof);
+
+  // 2. Verify Boot Proof
+  // Parse attestation
+  const coseSign1Der = Uint8Array.from(
+    atob(bootProof.awsAttestationDocB64)
+      .split("")
+      .map((c) => c.charCodeAt(0)),
+  );
+  const coseSign1 = CBOR.decode(coseSign1Der.buffer);
+  const [, , payload] = coseSign1;
+  const attestationDoc = CBOR.decode(new Uint8Array(payload).buffer);
+
+  // Verify cose sign1 signature
+  verifyCoseSign1Sig(coseSign1, attestationDoc.certificate);
+
+  // Verify certificate chain
+  const appProofTimestampMs = parseInt(
+    JSON.parse(appProof.proofPayload).timestampMs,
+  );
+  await verifyCertificateChain(
+    attestationDoc.cabundle,
+    AWS_ROOT_CERT_PEM,
+    attestationDoc.certificate,
+    appProofTimestampMs,
+  );
+
+  // Verify manifest digest
+  const decodedBootProofManifest = Uint8Array.from(
+    atob(bootProof.qosManifestB64)
+      .split("")
+      .map((c) => c.charCodeAt(0)),
+  );
+  const manifestHash = sha256(decodedBootProofManifest);
+  if (!bytesEq(manifestHash, attestationDoc.user_data)) {
+    throw new Error(
+      "attestationDoc's user_data doesn't match the hash of the manifest",
+    );
+  }
+
+  // 3. Verify that all the ephemeral public keys match: app proof, boot proof structure, actual attestation doc
+  const u8 = new Uint8Array(attestationDoc.public_key);
+  const attestationPubKey = Buffer.from(u8).toString("hex");
+  if (
+    appProof.publicKey !== attestationPubKey ||
+    attestationPubKey !== bootProof.ephemeralPublicKeyHex
+  ) {
+    throw new Error(
+      `Ephemeral pub keys from app proof: ${appProof.publicKey}, boot proof structure ${bootProof.ephemeralPublicKeyHex}, and attestation doc ${attestationPubKey} should all match`,
+    );
+  }
+}
+
+export function verifyAppProofSignature(appProof: v1AppProof): void {
+  if (!appProof) {
+    throw new Error("App proof cannot be null");
```

Comment:
nit: "cannot be falsy" (if `!appProof` is true, `appProof` can be `null`, `undefined`, the number 0, the empty string....and more. JS is [whacky](https://dorey.github.io/JavaScript-Equality-Table/))